crates/trie/sparse-parallel/src/trie.rs

```diff
@@ -175,23 +193,433 @@ impl ParallelSparseTrie {
+                // If only one child is left set in the branch node, we need to collapse it. Get
+                // full path of the only child node left.
+                let remaining_child_path = {
+                    let mut p = *branch_path;
+                    p.push_unchecked(
+                        state_mask.first_set_bit_index().expect("state mask is not empty"),
+                    );
+                    p
+                };
+
+                trace!(
+                    target: "trie::parallel_sparse",
+                    ?leaf_path,
+                    ?branch_path,
+                    ?remaining_child_path,
+                    "Branch node has only one child",
+                );
+
+                let remaining_child_subtrie = self.subtrie_for_path(&remaining_child_path);
+
+                // If the remaining child node is not yet revealed then we have to reveal it here,
+                // otherwise it's not possible to know how to collapse the branch.
+                let remaining_child_node = match remaining_child_subtrie
+                    .nodes
+                    .get(&remaining_child_path)
+                    .unwrap()
+                {
+                    SparseNode::Hash(_) => {
+                        trace!(
+                            target: "trie::parallel_sparse",
+                            ?remaining_child_path,
+                            "Retrieving remaining blinded branch child",
+                        );
+                        if let Some(RevealedNode { node, tree_mask, hash_mask }) =
+                            provider.blinded_node(&remaining_child_path)?
+                        {
+                            let decoded = TrieNode::decode(&mut &node[..])?;
+                            trace!(
+                                target: "trie::parallel_sparse",
+                                ?remaining_child_path,
+                                ?decoded,
+                                ?tree_mask,
+                                ?hash_mask,
+                                "Revealing remaining blinded branch child"
+                            );
+                            remaining_child_subtrie.reveal_node(
+                                remaining_child_path,
+                                &decoded,
+                                TrieMasks { hash_mask, tree_mask },
+                            )?;
+                            remaining_child_subtrie.nodes.get(&remaining_child_path).unwrap()
+                        } else {
+                            return Err(SparseTrieErrorKind::Other(Box::<
+                                dyn core::error::Error + Send + Sync,
+                            >::from(
+                                format!("remaining child node {remaining_child_path:?} not found in provider")
+                            ))
+                            .into())
+                        }
```

Comment:
in the regular sparse trie we don't return an error here, not sure that we need to here. But if we do, there should be an easier way to convert into the other error variant

Reply:
the new error variant makes sense